### PR TITLE
[IMP] udes_security: Port of redirection domain filter list

### PR DESCRIPTION
**Pull request by LoKKeR96**
Porting of the code that implements the allowlist domain filter model
for blocking redirections from UDES13. This code implements a list to
filter domains and block redirection requests if the domain is not in
the list filter. The list has been renamed from whitelist to allowlist.

Story: x2536

Signed-off-by: Lorenzo Cucurachi <lorenzo.cucurachi@unipart.io>

### CODE REVIEW

File: addons/udes_security/controllers/redirect.py [hunk 1]
```diff
@@ -1,11 +1,72 @@
...
+
+                if self.is_absolute_url(parsed_url) and not self.domain_is_allowlisted(parsed_url):
+                    _logger.error("Invalid redirection attempted: %s", redirect)
+                    raise BadRequest(_("Invalid redirection attempted"))
+
+        return self.func(*args, **kwargs)
+
+    @staticmethod
+    def is_absolute_url(parsed_url):
+        """Test if a URL is an absolute URL"""
```
Comment (PeterC10) [inline addons/udes_security/controllers/redirect.py:52]: Super nit: Remove the term `test` from the doc-string here and for `domain_is_allowlisted`.

E.g. `Returns True if URL is an absolute URL, else False` or similar.

File: addons/udes_security/data/res_users.xml [hunk 2]
```diff
@@ -13,6 +13,7 @@
                 (4, ref('udes_security.group_can_add_trusted_user')),
                 (4, ref('udes_security.group_trusted_user')),
                 (4, ref('udes_security.group_edit_allowed_file_types')),
+                (4, ref('udes_security.group_edit_allowlist_redirect_domains')),
```
Comment (PeterC10) [inline addons/udes_security/data/res_users.xml:16]: Can we add this to `user_root` as well please? Technically doesn't need them, but good to keep it consistent. :)

File: addons/udes_security/tests/test_no_redirect.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        )
+
+    def test_url_encoded_double_forward_slash_redirects_locally(self):
+        self.authenticate("admin", "admin")
+        self.assertNotFound(
+            self.url_open("/web?redirect=%2F%2Fbaddomain.com"),
+            redirect_url=self._get_url("/baddomain.com"),
+        )
+
+    def test_white_listed_domain_redirects(self):
```
Comment (PeterC10) [inline addons/udes_security/tests/test_no_redirect.py:84]: Supernit: `white` -> `allow`

File: addons/udes_security/tests/test_no_redirect.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+        self.assertFailedRedirect(self.url_open("/web?redirect=http://google.com"))
+
+        DomainAllowlist.create({"name": "Google", "domain": "google.com"})
+        self.assertSuccessfulRedirect(
+            self.url_open("/web?redirect=http://google.com"),
+            redirect_url="http://google.com",
+        )
+
+    def test_white_listed_wildcarded_domain_redirects(self):
```
[the remark this answers is not in the document]
Comment (LoKKeR96) [reply addons/udes_security/tests/test_no_redirect.py:97]: Can't trust the Ctrl+F Search anymore :) Good spot.

File: addons/udes_security/views/domain_allowlist_view.xml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        </tree>
+      </field>
+    </record>
+
+    <!-- Search filter -->
+    <record id="domain_allowlist_search" model="ir.ui.view">
+      <field name="name">udes_security.domain.allowlist.search</field>
+      <field name="model">udes_security.domain.allowlist</field>
+      <field name="arch" type="xml">
+        <search string="Allowlist Redirect Domains">
```
Comment (PeterC10) [inline addons/udes_security/views/domain_allowlist_view.xml:47]: Suggestion: Add a filter for archived records.

**Review by PeterC10 — APPROVED** [review]

**Review by PeterAlabaster — APPROVED** [review]

File: addons/udes_security/data/res_groups.xml [hunk 6]
```diff
@@ -41,4 +41,13 @@
         <field name="u_required_group_id_to_change" ref="base.group_system" />
     </record>
 
+    <record id="group_edit_allowlist_redirect_domains" model="res.groups">
+        <field name="name">Manage Allowed Redirect Domains</field>
+        <field name="category_id" ref="base.module_category_extra" />
+    </record>
+    <!-- Cannot self reference, so we extend the record after we create it -->
+    <record id="udes_security.group_edit_allowlist_redirect_domains" model="res.groups">
```
Comment (armandcela) [inline addons/udes_security/data/res_groups.xml:49]: Do we need to add the group in any of actual user templates, in order that users applied the specific template will have the right to modify allowed list.
I think is more a PO question

Comment (armandcela) [reply addons/udes_security/data/res_groups.xml:49]: Got the response i think is needed only for admin :)

**Review by armandcela — APPROVED** [review]

**Pull request merged**